docs/building-modules/invariants.md

```diff
@@ -1,5 +1,86 @@
+
+- [Implementing `Invariant`s](#implementing-invariants)
+- [Invariant Registry](#invariant-registry)
+
+## Implementing `Invariant`s
+
+An `Invariant` is a function that checks for a particular invariant within a module. Module `Invariant`s must follow the [`Invariant`s type](https://github.com/cosmos/cosmos-sdk/blob/master/types/invariant.go#L9):
+
+```go
+type Invariant func(ctx Context) (string, bool)
+```
+
+where the `string` return value is the invariant message, which can be used when printing logs, and the `bool` return value is the actual result of the invariant check. 
+
+In practice, each module implements `Invariant`s in a `internal/keeper/invariants.go` file within the module's folder. The standard is to implement one `Invariant` function per logical grouping of invariants with the following model:
+
+```go
+// Example for an Invariant that checks balance-related invariants
+
+func BalanceInvariants(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+        // Implement checks for balance-related invariants
+    }
+}
+```
+
+Additionally, module developers should generally implement an `AllInvariants` function that runs all the `Invariant`s functions of the module:
+
+```go
+// AllInvariants runs all invariants of the module.
+// In this example, the module implements two Invariants: BalanceInvariants and DepositsInvariants
+
+func AllInvariants(k Keeper) sdk.Invariant {
+
+	return func(ctx sdk.Context) (string, bool) {
+		res, stop := BalanceInvariants(k)(ctx)
+		if stop {
+			return res, stop
+		}
+
+		return DepositsInvariant(k)(ctx)
+	}
+}
+```
+
+Finally, module developers need to implement the `RegisterInvariants` method as part of the [`AppModule` interface](./module-manager.md#appmodule). Indeed, the `RegisterInvariants` method of the module, implemented in the `module.go` file, typically only defers the call to a `RegisterInvariants` method implemented in `internal/keeper/invariants.go`. The `RegisterInvariants` method registers a route for each `Invariant` function in the [`InvariantRegistry`](#invariant-registry):
+
+
+```go
+// RegisterInvariants registers all staking invariants
+func RegisterInvariants(ir sdk.InvariantRegistry, k Keeper) {
+	ir.RegisterRoute(types.ModuleName, "module-accounts",
+		BalanceInvariants(k))
+	ir.RegisterRoute(types.ModuleName, "nonnegative-power",
+		DepositsInvariant(k))
+}
+```
+
+For more, see an example of [`Invariant`s implementation from the `staking` module](https://github.com/cosmos/cosmos-sdk/blob/master/x/staking/keeper/invariants.go). 
+
+## Invariant Registry
+
+The `InvariantRegistry` is a registry where the `Invariant`s of all the modules of an application are registered. Typically, it is implemented as a specific module. The module must implement an object that follow the [`sdk.InvariantRegistry` interface](https://github.com/cosmos/cosmos-sdk/blob/master/types/invariant.go#L14-L17). 
```

Comment:
```suggestion
The `InvariantRegistry` is a registry where the `Invariant`s of all the modules of an application are registered. Each module must implement an object that follows the [`sdk.InvariantRegistry` interface](https://github.com/cosmos/cosmos-sdk/blob/master/types/invariant.go#L14-L17). 
```

This is confusing

Reply:
Hmm I see why you're confused. Maybe it shouldn't be in the `building-modules` section then. There is only one `InvariantRegistry` for the application, not one per module. This registry is implemented as a module (just like the `anteHandler`).

Reply:
Clarified